networking_nsxv3/api/rpc.py

```diff
@@ -231,69 +171,83 @@ def plugin(self):
+    def get_security_group(self, context, security_group_id):
+        id_rev = db.get_security_group_revision(context, security_group_id)
+        if id_rev:
+            return {
+                "id": id_rev[0],
+                "revision_number": id_rev[1],
+                "tags": db.get_security_group_tag(context, security_group_id),
+                "ports": db.get_port_id_by_sec_group_id(context, 
+                                                        security_group_id)
+            }
 
     @log_helpers.log_method_call
-    @oslo_messaging.expected_exceptions(exceptions.ObjectNotFound)
-    def get_qos(self, context, qos_id):
-        return db.get_qos(context, qos_id)
+    def get_rules_for_security_group_id(self, context, security_group_id):
+        return db.get_rules_for_security_group_id(context, security_group_id)
 
     @log_helpers.log_method_call
-    def get_qos_bwl_rules(self, context, qos_id):
-        return db.get_qos_bwl_rules(context, qos_id)
+    def get_security_group_members_effective_ips(self, context, security_group_id):
+        return \
+            db.get_security_group_members_ips(context, security_group_id) + \
+            db.get_security_group_members_address_bindings_ips(context, security_group_id)
 
     @log_helpers.log_method_call
-    def get_qos_dscp_rules(self, context, qos_id):
-        return db.get_qos_dscp_rules(context, qos_id)
+    def get_security_groups_for_host(self, context, host, limit, cursor):
+        return db.get_security_groups_for_host(context, host, limit, cursor)
 
     @log_helpers.log_method_call
-    @oslo_messaging.expected_exceptions(exceptions.ObjectNotFound)
-    def get_port(self, context, port_id):
-        return db.get_port(context, port_id)
-
+    def get_remote_security_groups_for_host(self, context, host, limit, cursor):
+        return db.get_remote_security_groups_for_host(context, host, limit, cursor)
+    
     @log_helpers.log_method_call
-    def get_port_security_groups(self, context, port_id):
-        return db.get_port_security_groups(context, port_id)
+    def has_security_group_used_by_host(self, context, host, security_group_id):
+        return db.has_security_group_used_by_host(context, host, security_group_id)
 
     @log_helpers.log_method_call
-    def get_port_allowed_pairs(self, context, port_id):
-        return db.get_port_allowed_pairs(context, port_id)
+    def get_port(self, context, host, port_id):
+        port = db.get_port(context, host, port_id)
 
-    @log_helpers.log_method_call
-    def get_port_addresses(self, context, port_id):
-        return db.get_port_addresses(context, port_id)
+        # NSX-T does not support CIDR as port manual binding - skipping X/X
 
-    @log_helpers.log_method_call
-    def get_rules_for_security_groups_id(self, context, security_group_id):
-        return db.get_rules_for_security_groups_id(context, security_group_id)
+        for ip in db.get_port_addresses(context, port_id):
```

Comment:
get_port_addresses() returns a list of [ip_address, subnet_id], which causes later in the Ports call an invalid structure.

If the subnet_id is not needed, maybe completely remove from rpc?

Reply:
e.g. 
```
2021-05-21 17:26:30,872.872 86289 ERROR networking_nsxv3.plugins.ml2.drivers.nsxv3.agent.client_nsx [-] Request=Function PUT Argumetns {'path': '/api/v1/logical-ports/a4a2d2f3-c9cd-4780-92f1-4f1f5ed54347', 'data': {'switching_profile_ids': [{'key': 'SpoofGuardSwitchingProfile', 'value': u'f833742f-459f-4783-b20c-9bd61752675c'}, {'key': 'IpDiscoverySwitchingProfile', 'value': u'b73982a9-c0ea-4632-a62f-d4b831ac4598'}], 'tags': [{'scope': 'security_group', 'tag': u'4360da8e-b208-4bcf-8021-7a773d795f84'}, {'scope': 'age', 'tag': 1621610790}, {'scope': 'agent_id', 'tag': 'nsxm-l-01a.corp.local'}, {'scope': 'revision_number', 'tag': 10}], 'address_bindings': [{u'ip_address': [u'10.180.0.17', u'858f0d2d-5622-4be2-abe6-e5a286bde3b7'], u'mac_address': u'fa:16:3e:49:44:ae'}], 'logical_switch_id': u'91d1e26c-7c03-493b-951f-b7f88f6fa6eb', 'admin_state': 'UP', 'attachment': {'attachment_type': 'VIF', 'context': {'vif_type': 'PARENT', 'traffic_tag': 2054, 'resource_type': 'VifAttachmentContext'}, 'id': u'e46fc1b8-27bd-4353-bd44-b18f57df6baa'}, '_revision': 194, 'display_name': u'e46fc1b8-27bd-4353-bd44-b18f57df6baa'}} Response=Error Code=400 Message={
  "httpStatus" : "BAD_REQUEST",
  "error_code" : 208,
  "module_name" : "common-services",
  "error_message" : "The property address_bindings.null.ip_address must be a valid string"
}
```

Reply:
Fixed